File: src/main/java/org/xodium/vanillaplus/VanillaPlus.java
package org.xodium.vanillaplus;

import org.bukkit.plugin.java.JavaPlugin;
import org.xodium.vanillaplus.commands.ReloadCommand;

import java.util.Arrays;

public class VanillaPlus extends JavaPlugin {
    // TODO: revert back to using the build in config system instead of a database.

    private static final String[] V = { "1.21.1", "1.21.3", "1.21.4" };
    private static final String[] PAPER = { "Paper" };
    private static final String IS_PAPER_MSG = "This plugin is not compatible with non-Paper servers.";
    private static final String IS_SUPPORTED_VERSION_MSG = "This plugin requires Paper version(s): "
            + String.join(", ", V);

    public static final String PREFIX = "<gold>[<dark_aqua>VanillaPlus<gold>] <reset>";

    public static VanillaPlus getInstance() {
        return getPlugin(VanillaPlus.class);
    }

    @Override
    public void onEnable() {
        if (!isPaper()) {
            disablePlugin(IS_PAPER_MSG);
            return;
        }
        if (!isSupportedVersion()) {
            disablePlugin(IS_SUPPORTED_VERSION_MSG);
            return;
        }
        saveDefaultConfig();
        new ReloadCommand();
        new ModuleManager();
    }

    private void disablePlugin(String msg) {
        getLogger().severe(msg);
        getServer().getPluginManager().disablePlugin(this);
    }

    private boolean isSupportedVersion() {
        return Arrays.stream(V)
                .anyMatch(v -> getServer().getVersion().contains(v));
    }

    private boolean isPaper() {
        return Arrays.stream(PAPER)
                .anyMatch(v -> getServer().getName().contains(v));
    }
}

File: src/main/kotlin/org/xodium/vanillaplus/VanillaPlus.kt
@file:Suppress("ktlint:standard:no-wildcard-imports")

package org.xodium.vanillaplus

import io.papermc.paper.plugin.lifecycle.event.types.LifecycleEvents
import org.bukkit.plugin.java.JavaPlugin
import org.xodium.vanillaplus.data.ConfigData
import org.xodium.vanillaplus.data.ConfigData.Companion.load
import org.xodium.vanillaplus.modules.*
import org.xodium.vanillaplus.recipes.ChainmailRecipe
import org.xodium.vanillaplus.recipes.DiamondRecycleRecipe
import org.xodium.vanillaplus.recipes.PaintingRecipe
import org.xodium.vanillaplus.recipes.RottenFleshRecipe
import org.xodium.vanillaplus.recipes.WoodLogRecipe

/** Main class of the plugin. */
internal class VanillaPlus : JavaPlugin() {
    companion object {
        lateinit var instance: VanillaPlus
            private set

        lateinit var configData: ConfigData
    }

    init {
        instance = this
    }

    override fun onEnable() {
        val unsupportedVersionMsg =
            "This plugin requires a supported server version. Supported versions: ${pluginMeta.version}."

        if (!server.version.contains(pluginMeta.version.substringBefore("+"))) disablePlugin(unsupportedVersionMsg)

        instance.lifecycleManager.registerEventHandler(LifecycleEvents.COMMANDS) {
            it.registrar().register(
                ConfigData.reloadCommand.builder.build(),
                ConfigData.reloadCommand.description,
                ConfigData.reloadCommand.aliases,
            )
        }

        instance.server.pluginManager.addPermission(ConfigData.reloadPermission)

        configData = ConfigData().load("config.json")

        val recipes =
            listOf(
                ChainmailRecipe,
                DiamondRecycleRecipe,
                PaintingRecipe,
                RottenFleshRecipe,
                WoodLogRecipe,
            )

        logger.info(
            "Registered: ${recipes.sumOf { it.recipes.size }} recipes(s) | Took ${recipes.sumOf { it.register() }}ms",
        )

        val modules =
            listOf(
                BookshelfModule,
                BooksModule,
                ChatModule,
                DimensionsModule,
                EntityModule,
                InventoryModule,
                LocatorModule,
                MannequinModule,
                MapModule,
                MotdModule,
                OpenableModule,
                PlayerModule,
                ServerInfoModule,
                ScoreBoardModule,
                SitModule,
                TabListModule,
                TameableModule,
            ).filter { it.isEnabled }

        logger.info(
            "Registered: ${modules.size} module(s) | Took ${modules.sumOf { it.register() }}ms",
        )
    }

    override fun onDisable() {
        server.messenger.unregisterOutgoingPluginChannel(instance)
    }

    /**
     * Disable the plugin and log the message.
     * @param msg The message to log.
     */
    private fun disablePlugin(msg: String): Nothing {
        logger.severe(msg)
        server.pluginManager.disablePlugin(instance)
        throw IllegalStateException(msg)
    }
}
